File: number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        List<Integer>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            int a = edge[0];
            int b = edge[1];
            
            adj[a].add(b);
            adj[b].add(a);
        }
        
        boolean[] seen = new boolean[n];
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (!seen[i]) {
                count++;
                
                Queue<Integer> q = new LinkedList<>();
                q.add(i);
                seen[i] = true;
                
                while(!q.isEmpty()) {
                    int node = q.poll();
                    
                    for (int neighbor : adj[node]) {
                        if (!seen[neighbor]) {
                            seen[neighbor] = true;
                            q.add(neighbor);
                        }
                    }
                }
            }
        }
        
        return count;
    }
}

File: number-of-connected-components-in-an-undirected-graph/number-of-connected-components-in-an-undirected-graph.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        
        adj = [[] for _ in range(n)]
        
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)
        
        visited = [False for _ in range(n)]
        
        count = 0
        
        for i in range(n):
            if not visited[i]:
                count += 1
                stack = [i]
                visited[i] = True
                
                while stack:
                    node = stack.pop()
                    for neighbor in adj[node]:
                        if not visited[neighbor]:
                            stack.append(neighbor)
                            visited[neighbor] = True
                
        
        return count
        